File: react-native/react-native-totvs-clockin-vision/ios/Events/OnFaceRecognized.swift
//
//  OnFaceRecognized.swift
//  TOTVSClockInVision
//
//  Created by Jansel Rodriguez on 8/3/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation


fileprivate let FIELD_PERSON_ID = "personid"
fileprivate let FIELD_PERSON_NAME = "name"
fileprivate let FIELD_CONFIDENCE = "confidence"
fileprivate let FIELD_IMAGE_ENCODING = "imageEncoding"
fileprivate let FIELD_IMAGE_PATH = "imageFilePath"
fileprivate let FIELD_RESULTS = "results"
fileprivate let FIELD_STATUS = "status"

/**
* Event emitted when face recognition is triggered.
*/
public class OnFaceRecognized : Event {
    
    private let emit: RCTDirectEventBlock?
    
    public required init(emit: RCTDirectEventBlock?) {
        self.emit = emit
    }
    
    public func send(data: RecognitionResult) {
        var persons = [[String: Any]]()
        for face in data.output.entities {
            if face.personId.isEmpty {
                continue
            }
            persons.append([
                FIELD_PERSON_ID: face.personId,
                FIELD_PERSON_NAME: face.name,
                FIELD_CONFIDENCE: face.distance,
                FIELD_IMAGE_ENCODING: data.output.encoding
            ])
        }
        var event: [String: Any] = [
            FIELD_RESULTS: persons,
            FIELD_STATUS: data.output.status
        ]
        if nil != data.imagePath {
            event[FIELD_IMAGE_PATH] = data.imagePath!.path
        }
        emit?(event)
    }
}

File: react-native/react-native-clockin-vision/android/src/main/java/com/totvs/clockin/vision/events/OnFaceRecognized.kt
package com.totvs.clockin.vision.events

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.uimanager.events.RCTEventEmitter
import com.totvs.clockin.vision.core.ModelOutput
import com.totvs.clockin.vision.face.Face
import java.io.File

/**
 * Event emitted when face recognition is triggered.
 */
class OnFaceRecognized(
    private val savedImage: File? = null,
    private val output: ModelOutput<Face>
) : Event {
    override fun invoke(context: ReactContext, viewId: Int) {
        val persons = Arguments.createArray()
        for (face in output.entities) {
            if (face.personId.isEmpty()) {
                continue
            }
            val person = Arguments.createMap().apply {
                putString(FIELD_PERSON_ID, face.personId)
                putString(FIELD_PERSON_NAME, face.name)
                putDouble(FIELD_CONFIDENCE, face.distance.toDouble())
                putString(FIELD_IMAGE_ENCODING, output.encoding)
            }
            persons.pushMap(person)
        }

        val event = Arguments.createMap().apply {
            savedImage?.let { path ->
                putString(FIELD_IMAGE_PATH, path.absolutePath)
            }
            putString(FIELD_STATUS, output.status)
            putArray(FIELD_RESULTS, persons)
        }
        // dispatch the event.
        context.getJSModule(RCTEventEmitter::class.java).receiveEvent(viewId, name, event)
    }

    companion object : Event.Export {
        override val name: String = "onFaceRecognized"

        private const val FIELD_PERSON_ID = "personid"
        private const val FIELD_PERSON_NAME = "name"
        private const val FIELD_CONFIDENCE = "confidence"
        private const val FIELD_IMAGE_ENCODING = "imageEncoding"
        private const val FIELD_IMAGE_PATH = "imageFilePath"
        private const val FIELD_RESULTS = "results"
        private const val FIELD_STATUS = "status"
    }
}